File: problem-2.js
var limit = 4000000;

//generates the gibonacci sequence up to the given limit.
var genFib = function(limit){
	var fib = [1,2,3];
	var sum = 0;
	while(sum < limit){
		sum = fib[fib.length - 1] + fib[fib.length - 2];
		fib.push(sum);
	};
	return fib;
};

//returns an array of the even elements of the given array.
var evenArray = function(a){
	var even = [];
	for(var i = 0; i < a.length; i++){
		if(a[i] % 2 == 0){
			even.push(a[i]);
		};
	};
	return even;
};

//returns the sum of an array
var sumArray = function(a){
	var sum = 0;
	for(var i = 0; i<a.length; i++){
		sum += a[i];
	};
	return sum;
};

print(sumArray(evenArray(genFib(limit))));

File: problem-2.rb
max = 4000000

#returns an array of the fibonacci sequence where the
#final digit is less than or equal to the given max
def generate_fibonacci max
	fib = [1,2,3]
	while fib[-1] <= max do
		fib<<fib[-1]+fib[-2]
	end
	fib
end

#returns an array of the even elements from the given array
def find_evens array
	evens = []
	array.each {|num| evens<<num if num % 2 == 0}
	evens
end

#returns the sum of the elements in the given array
def sum_array array
	sum = 0
	array.each {|n| sum += n}
	sum
end

puts sum_array(find_evens(generate_fibonacci(max)))